File: typescript/src/ui/overlays/starfield/starfield.ts
import P5 from 'p5';
import Star from './star';

const STARS_COUNT = 400;

export default class Starfield {
    p5: P5;
    stars: Star[] = [];

    constructor(p5: P5) {
        this.p5 = p5;

        for (let counter = 0; counter < STARS_COUNT; counter++) {
            this.stars.push(new Star(p5));
        }
    }

    update() {
        for (let star of this.stars) {
            star.update();
        }
    }

    draw() {
        for (let star of this.stars) {
            this.p5.push();

            this.p5.translate(this.p5.width / 2, this.p5.height / 2);

            star.draw();

            this.p5.pop();
        }
    }
}

File: rust/src/overlays/starfield.rs
use web_sys::CanvasRenderingContext2d;

use super::star::Star;
use crate::{sprites::sprite::CanvasDimension, utils::config::Config};

pub struct Starfield {
    stars: Vec<Star>,
    canvas: CanvasDimension,
}

impl Starfield {
    pub fn new(canvas: CanvasDimension) -> Starfield {
        let config = Config::new();

        let mut stars = Vec::<Star>::new();

        for _counter in 0..config.overlays.starfield.stars_count {
            stars.push(Star::new(canvas));
        }

        Starfield { stars, canvas }
    }
}

impl Starfield {
    pub fn update(&mut self) {
        for star in &mut self.stars {
            star.update();
        }
    }

    pub fn draw(&mut self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        let _result = canvas.translate(self.canvas.width / 2.0, self.canvas.height / 2.0);

        for star in &mut self.stars {
            star.draw(canvas.clone());
        }

        canvas.restore();
    }
}
